iothub/device/src/Transport/Amqp/AmqpUnit.cs

```diff
@@ -396,50 +445,58 @@ public async Task<AmqpIoTOutcome> SendMethodResponseAsync(MethodResponseInternal
-                if (_twinLinksOpened) return;
+                throw DEVICE_DISCONNECTED_EXCEPTION;
+            }
 
-                Debug.Assert(_twinSendingLink == null);
-                Debug.Assert(_twinReceivingLink == null);
+            if (Logging.IsEnabled) Logging.Enter(this, timeout, $"{nameof(EnableTwinLinksAsync)}");
+            AmqpIoTSession amqpIoTSession = await EnsureSessionAsync(timeout).ConfigureAwait(false);
+            bool gain = await _twinLinksLock.WaitAsync(timeout).ConfigureAwait(false);
+            if (!gain)
+            {
+                throw new TimeoutException();
+            }
 
+            try
+            {
                 string correlationIdSuffix = Guid.NewGuid().ToString();
 
-                Task<AmqpIoTReceivingLink> receiveLinkCreator = _amqpIoTSession.OpenTwinReceiverLinkAsync(_deviceIdentity, correlationIdSuffix, timeout);
-                Task<AmqpIoTSendingLink> sendingLinkCreator = _amqpIoTSession.OpenTwinSenderLinkAsync(_deviceIdentity, correlationIdSuffix, timeout);
-                await Task.WhenAll(receiveLinkCreator, sendingLinkCreator).ConfigureAwait(false);
-
-                _twinSendingLink = sendingLinkCreator.Result;
-                _twinSendingLink.Closed += OnLinkDisconnected;
-
-                _twinReceivingLink = receiveLinkCreator.Result;
-                _twinReceivingLink.RegisterTwinListener(OnDesiredPropertyReceived);
-                _twinReceivingLink.Closed += OnLinkDisconnected;
-
-                _twinLinksOpened = true;
-
-                if (Logging.IsEnabled) Logging.Associate(this, this, _twinReceivingLink, $"{nameof(EnsureTwinLinksAreOpenedAsync)}");
-                if (Logging.IsEnabled) Logging.Associate(this, this, _twinSendingLink, $"{nameof(EnsureTwinLinksAreOpenedAsync)}");
+                await Task.WhenAll(
+                   OpenTwinReceiverLinkAsync(amqpIoTSession, correlationIdSuffix, timeout),
+                   OpenTwinSenderLinkAsync(amqpIoTSession, correlationIdSuffix, timeout)
+               ).ConfigureAwait(false);
             }
-            catch (Exception ex) when (!ex.IsFatal())
+            finally
             {
-                _twinReceivingLink?.Abort();
-                _twinSendingLink?.Abort();
-                _twinReceivingLink = null;
-                _twinSendingLink = null;
+                _twinLinksLock.Release();
+                if (Logging.IsEnabled) Logging.Exit(this, timeout, $"{nameof(EnableTwinLinksAsync)}");
+            }
+        }
 
-                throw;
+        private async Task OpenTwinReceiverLinkAsync(AmqpIoTSession amqpIoTSession, string correlationIdSuffix, TimeSpan timeout)
+        {
+            if (_twinReceivingLink == null || _twinReceivingLink.IsClosing())
+            {
+                _twinReceivingLink = await amqpIoTSession.OpenTwinReceiverLinkAsync(_deviceIdentity, correlationIdSuffix, timeout).ConfigureAwait(false);
```

Comment:
Same as above: this has a race condition.

Reply:
Offline discussion. Table P0